### DISCUSSION TITLE
Components `Transitive` attribute

### DISCUSSION THREAD

**Discussion by l0rd** [wixtoolset/issues#8617] [Questions] [answered]
We are using a `Component` to create a configuration file under `C:\ProgramData\a\b\c\d.conf`
Under some circumstances we don't want that file to be created and, for that, we use a `Condition`.
This is a simplified version of the definition of the component in the WXS file:

```xml
<StandardDirectory Id="CommonAppDataFolder">
  <Directory Id="..." Name="a">
    <Directory Id="..." Name="b">
      <Component Id="..." Guid="..." Condition="...">
        <CreateFolder />
        <IniFile Id="..." Action="createLine" Directory="c" Section="..." Name="d.conf" Key="..." Value="..." />
      </Component>
    </Directory>
  </Directory>
</StandardDirectory>
```

All works well on first installation: the file `C:\ProgramData\a\b\c\d.conf` is created when the condition is satisfied.

But when updating, if the condition is not satisfied, the file `C:\ProgramData\a\b\c\d.conf` gets deleted and we want to avoid that.

[Reading the Components attributes documentation](https://wixtoolset.org/docs/schema/wxs/component/#attributes) we found `Transitive` that seams to do what we are looking for. But after setting `Transitive="false"` the behavior is still the same:

```xml
...
      <Component Id="..." Guid="..." Condition="..." Transitive="false">
      ...
      </Component>
...
```

Do you have any advice to use `Transient`? And in general, how can we avoid deleting a component on subsequent installations, even if the condition is not met anymore?

**Comment by bevanweiss** [comment]
You should probably consult the Microsoft installer component information around this.
https://learn.microsoft.com/en-us/windows/win32/msi/windows-installer-components

Also note, setting Transitive='false' is the default.  You don't need to include the attribute if you want the 'false' behaviour.. that's what you are getting.

It sounds like you might want Permanent='true' for this file component, and then handle removal in a different manner.

**Comment by l0rd** [comment]
Thank you @bevanweiss. I have tried using `Permanent="true"` and that works. And I am also ok with the file not being removed when uninstalling. But having "an extra system client for the component in the Windows Installer registry settings", that is kept as an orphan after uninstall, worries me more. Do you have any suggestion to clean up every trace from the Windows installer registry settings?

**Comment by l0rd** [comment, marked as answer]
I have addressed this problem by setting the `Condition` to true when the INI file already exists and setting `NeverOverwrite="true"`.

**Reply by bevanweiss** [reply]
My expectation is that on an uninstall, this file would still get removed.  Is this what you intend?
Often these 'user configuration' items would be desired to remain (unless the user has specifically chosen to remove them on uninstall) so that if they are intending to install again the file is still in place.
It can often be the case that it's just 'easier' to handle user configuration files outside of the installer framework (so the application itself creates the initial 'template' config files if they don't already exist), and the installer just has conditional RemoveFiles if the user has chosen to remove configuration on uninstall.

**Reply by l0rd** [reply]
Yes, the idea is to remove the configuration file on an uninstall. The choice is for the uninstaller to be responsible for the deletion of every object that it has created. Leaving the configuration file is an acceptable approach in my opinion, but the deletion should be preferred and that has motivated my choice to avoid `Permanent="true"`.
